### PR TITLE
BLADE-361 Blade Deploy rewrite

### PR DESCRIPTION
**Pull request by christopherbryanboyd**
Hi @gamerson , this addresses https://issues.liferay.com/browse/BLADE-361

### CODE REVIEW

File: cli/src/main/java/com/liferay/blade/cli/BladeCLI.java [hunk 1]
```diff
@@ -262,6 +262,14 @@ public void run(String[] args) throws Exception {
 
 		BladeSettings bladeSettings = getBladeSettings();
 
+		if (!WorkspaceUtil.isWorkspace(this)) {
```
Comment (gamerson) [inline cli/src/main/java/com/liferay/blade/cli/BladeCLI.java]: This should go to maven-profile.  Can we do this in a separate PR?

File: cli/src/main/java/com/liferay/blade/cli/command/CreateCommand.java [hunk 2]
```diff
@@ -195,12 +195,24 @@ else if (template.startsWith("modules-ext")) {
 			return;
 		}
 
+		boolean mavenBuild = "maven".equals(createArgs.getBuild());
```
Comment (gamerson) [inline cli/src/main/java/com/liferay/blade/cli/command/CreateCommand.java]: move to maven-profile

File: cli/src/main/java/com/liferay/blade/cli/util/WorkspaceUtil.java [hunk 3]
```diff
@@ -72,7 +73,9 @@ public static File getWorkspaceDir(File dir) {
 			return gradleParent;
 		}
 
-		File mavenParent = BladeUtil.findParentFile(dir, new String[] {"pom.xml"}, true);
+		Predicate<File> isWorkspacePomFile = WorkspaceUtil::_isWorkspacePomFile;
```
Comment (gamerson) [inline cli/src/main/java/com/liferay/blade/cli/util/WorkspaceUtil.java]: Need to add to maven-profile

File: gradle-tooling/src/main/java/com/liferay/blade/gradle/tooling/DefaultModel.java [hunk 4]
```diff
@@ -27,9 +27,15 @@
  */
 public class DefaultModel implements ProjectInfo, Serializable {
 
-	public DefaultModel(Set<String> pluginClassNames, Map<String, Set<File>> projectOutputFiles) {
+	public DefaultModel(Set<String> pluginClassNames, Map<String, Set<File>> projectOutputFiles, String liferayHome) {
```
Comment (gamerson) [inline gradle-tooling/src/main/java/com/liferay/blade/gradle/tooling/DefaultModel.java]: We need to bump the version of the tooling model api, to 1.1.0-SNAPSHOT

**Comment by christopherbryanboyd** [comment]
Hi @gamerson , I have removed the changes from here and will include them in the other PR specifically for the maven profile

File: cli/src/main/java/com/liferay/blade/cli/command/CreateCommand.java [hunk 5]
```diff
@@ -251,6 +232,50 @@ protected void execute(ProjectTemplatesArgs projectTemplatesArgs) throws Excepti
 		}
 	}
 
+	protected String getLiferayVersion(BladeCLI bladeCLI, CreateArgs createArgs) throws IOException {
```
Comment (gamerson) [inline cli/src/main/java/com/liferay/blade/cli/command/CreateCommand.java]: Lets remove this protected method and just make a private copy in profiles.

File: cli/src/main/java/com/liferay/blade/cli/command/DeployCommand.java [hunk 6]
```diff
@@ -106,297 +71,121 @@ public void execute() throws Exception {
 		return DeployArgs.class;
 	}
 
+	protected boolean isWorkspace(File file) {
```
Comment (gamerson) [inline cli/src/main/java/com/liferay/blade/cli/command/DeployCommand.java]: Lets move this to private.

File: gradle-tooling/src/main/java/com/liferay/blade/gradle/tooling/ProjectInfo.java [hunk 7]
```diff
@@ -26,6 +26,10 @@
  */
 public interface ProjectInfo {
 
+	public default String getLiferayHome() {
```
Comment (gamerson) [inline gradle-tooling/src/main/java/com/liferay/blade/gradle/tooling/ProjectInfo.java]: Lets remove this default method for now.

File: gradle-tooling/src/main/java/com/liferay/blade/gradle/tooling/ProjectInfoPlugin.java [hunk 8]
```diff
@@ -111,14 +120,48 @@ public Object buildAll(String modelName, Project project) {
...
 
+		private String _getLiferayHome(Project project) {
+			ExtensionContainer extensionContainer = project.getExtensions();
+
+			Object liferayExtension = extensionContainer.findByName("liferay");
+
+			String liferayHome = null;
+
+			if (liferayExtension != null) {
+				Class<?> aclass = liferayExtension.getClass();
```
Comment (gamerson) [inline gradle-tooling/src/main/java/com/liferay/blade/gradle/tooling/ProjectInfoPlugin.java]:
```suggestion
				Class<?> clazz = liferayExtension.getClass();
```

File: gradle-tooling/src/main/java/com/liferay/blade/gradle/tooling/ProjectInfoPlugin.java [hunk 9]
```diff
@@ -111,14 +120,48 @@ public Object buildAll(String modelName, Project project) {
...
+
+				try {
+					BeanInfo beanInfo = Introspector.getBeanInfo(aclass);
+
+					for (PropertyDescriptor pd : beanInfo.getPropertyDescriptors()) {
```
Comment (gamerson) [inline gradle-tooling/src/main/java/com/liferay/blade/gradle/tooling/ProjectInfoPlugin.java]:
```suggestion
					for (PropertyDescriptor propertyDescriptor : beanInfo.getPropertyDescriptors()) {
```

**Comment by gamerson** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by christopherbryanboyd** [comment]
ci:retest

**Comment by gamerson** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

File: cli/src/main/java/com/liferay/blade/cli/util/BladeUtil.java [hunk 10]
```diff
@@ -90,7 +90,10 @@ public static void downloadLink(String link, Path target) throws IOException {
 		}
 	}
 
-	public static File findParentFile(File dir, String[] fileNames, boolean checkParents) {
+	@SafeVarargs
+	public static File findParentFile(
+		File dir, String[] fileNames, boolean checkParents, Predicate<File>... predicates) {
```
Comment (gamerson) [inline cli/src/main/java/com/liferay/blade/cli/util/BladeUtil.java]: THere are no callers that use the `predicates` argument.  Can you remove it?  THanks.

**Comment by gamerson** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by gamerson** [comment]
Pull request submitted to https://github.com/gamerson/liferay-blade-cli/pull/686. [See changes here.](https://github.com/gamerson/liferay-blade-cli/compare/e26811c9~1...e26811c9) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*
